### ISSUE TITLE
Role setPosition updating weird

### ISSUE THREAD

**Issue by firynth** [hydrabolt/discord.js#1158]
After calling `guild.setRolePosition()` a few times, it will push the roles underneath the specified position to have a position of 0 eventually. If used in such a way to dynamically add a role based on the position of another role, this can cause that role, and several roles beneath it to be pushed to 0 until Discord or the API sorts itself out and then assigns new positions to the roles, sometimes in a different order.

**Comment by GusCaplan** [comment]
this is intended behaviour

**Comment by firynth** [comment]
Then it needs to not be intended anymore because it's actively causing issues on the server where I have roles being set up. Five roles + the @everyone role cannot all have the position of 0 on a server.

**Comment by GusCaplan** [comment]
this behavior is intended by discord, not us. they order them based on id

**Comment by firynth** [comment]
No, I submitted another issue before and someone coded something to stop the position numbers from being inflated. I'm assuming that this is what is causing this issue because I went from version 10.0.4 or whatever to 11 and it started happening.

**Comment by GusCaplan** [comment]
this is, in fact, not a library issue.
![](https://i.gus.host/pRg5hx)
here we see four roles created one after another and not reordered in any way.
they all have the same position, and are ordered by id.

**Comment by firynth** [comment]
The issue is not with creation, is with setRolePosition. If they were numbered the way they should be, creating a new role, then setting the position dynamically based on the position of a correctly numbered role is what's causing the problem.

If you set it based on the role with position 4 because you want it inserted at that position, it and all roles with lower positions are reduced by one. The new dynamic set moves to 3 now and if you create another role and set position dynamically, that role and all roles below once again are reduced by one. Rinse and repeat until things break.

**Comment by GusCaplan** [comment]
still intended behavior from discord.

if i must spoonfeed you;
```js
getSortedRoles(guild) {
  return guild.roles.array()
    .sort((r1, r2) => (r1.position !== r2.position) ? r2.position - r1.position : r1.id - r2.id);
}
```

**Comment by firynth** [comment]
I just duplicated the code in src/structures/Guild.js and tested it with this

```javascript
var roles = new Array();
	    for(var i = 0; i < 20; i++) {
	        roles.push({"id" : i, "position" : i});
	    }
	    
	    roles.push({"id" : 20, "position" : 1});
	    var role = {"id" : 20, "position" : 1};
	    
	    roles = testFunction(roles, role, 7);
	    roles.push({"id" : 21, "position" : 1});
        role = {"id" : 21, "position" : 1};
        roles = testFunction(roles, role, 7);
        roles.push({"id" : 22, "position" : 1});
        role = {"id" : 22, "position" : 1};
        roles = testFunction(roles, role, 7);
        roles.push({"id" : 23, "position" : 1});
        role = {"id" : 23, "position" : 1};
        roles = testFunction(roles, role, 7);
        roles.push({"id" : 24, "position" : 1});
        role = {"id" : 24, "position" : 1};
        roles = testFunction(roles, role, 7);
```

This is the god damn output
`[ { id: 0, position: 0 },
  { id: 1, position: 0 },
  { id: 2, position: 1 },
  { id: 3, position: 2 },
  { id: 4, position: 3 },
  { id: 5, position: 4 },
  { id: 6, position: 5 },
  { id: 7, position: 6 },
  { id: 8, position: 8 },
  { id: 9, position: 9 },
  { id: 10, position: 10 },
  { id: 11, position: 11 },
  { id: 12, position: 12 },
  { id: 13, position: 13 },
  { id: 14, position: 14 },
  { id: 15, position: 15 },
  { id: 16, position: 16 },
  { id: 17, position: 17 },
  { id: 18, position: 18 },
  { id: 19, position: 19 },
  { id: 20, position: 7 } ]
[ { id: 0, position: 0 },
  { id: 1, position: 0 },
  { id: 2, position: 0 },
  { id: 3, position: 1 },
  { id: 4, position: 2 },
  { id: 5, position: 3 },
  { id: 6, position: 4 },
  { id: 7, position: 5 },
  { id: 8, position: 8 },
  { id: 9, position: 9 },
  { id: 10, position: 10 },
  { id: 11, position: 11 },
  { id: 12, position: 12 },
  { id: 13, position: 13 },
  { id: 14, position: 14 },
  { id: 15, position: 15 },
  { id: 16, position: 16 },
  { id: 17, position: 17 },
  { id: 18, position: 18 },
  { id: 19, position: 19 },
  { id: 20, position: 6 },
  { id: 21, position: 7 } ]
[ { id: 0, position: 0 },
  { id: 1, position: 0 },
  { id: 2, position: 0 },
  { id: 3, position: 0 },
  { id: 4, position: 1 },
  { id: 5, position: 2 },
  { id: 6, position: 3 },
  { id: 7, position: 4 },
  { id: 8, position: 8 },
  { id: 9, position: 9 },
  { id: 10, position: 10 },
  { id: 11, position: 11 },
  { id: 12, position: 12 },
  { id: 13, position: 13 },
  { id: 14, position: 14 },
  { id: 15, position: 15 },
  { id: 16, position: 16 },
  { id: 17, position: 17 },
  { id: 18, position: 18 },
  { id: 19, position: 19 },
  { id: 20, position: 5 },
  { id: 21, position: 6 },
  { id: 22, position: 7 } ]
[ { id: 0, position: 0 },
  { id: 1, position: 0 },
  { id: 2, position: 0 },
  { id: 3, position: 0 },
  { id: 4, position: 0 },
  { id: 5, position: 1 },
  { id: 6, position: 2 },
  { id: 7, position: 3 },
  { id: 8, position: 8 },
  { id: 9, position: 9 },
  { id: 10, position: 10 },
  { id: 11, position: 11 },
  { id: 12, position: 12 },
  { id: 13, position: 13 },
  { id: 14, position: 14 },
  { id: 15, position: 15 },
  { id: 16, position: 16 },
  { id: 17, position: 17 },
  { id: 18, position: 18 },
  { id: 19, position: 19 },
  { id: 20, position: 4 },
  { id: 21, position: 5 },
  { id: 22, position: 6 },
  { id: 23, position: 7 } ]
[ { id: 0, position: 0 },
  { id: 1, position: 0 },
  { id: 2, position: 0 },
  { id: 3, position: 0 },
  { id: 4, position: 0 },
  { id: 5, position: 0 },
  { id: 6, position: 1 },
  { id: 7, position: 2 },
  { id: 8, position: 8 },
  { id: 9, position: 9 },
  { id: 10, position: 10 },
  { id: 11, position: 11 },
  { id: 12, position: 12 },
  { id: 13, position: 13 },
  { id: 14, position: 14 },
  { id: 15, position: 15 },
  { id: 16, position: 16 },
  { id: 17, position: 17 },
  { id: 18, position: 18 },
  { id: 19, position: 19 },
  { id: 20, position: 3 },
  { id: 21, position: 4 },
  { id: 22, position: 5 },
  { id: 23, position: 6 },
  { id: 24, position: 7 } ]`

It's the code in your guy's library that's doing it.

**Comment by GusCaplan** [comment]
this isn't actually an issue, but you can consider it "fixed" in #1159
